File: 2015/02/rust/src/main.rs
use regex::Regex;
use std::cmp::{ max, min };
use std::fs::File;
use std::io::{ BufRead, BufReader };

// --- Constants --
const INPUT_FILE: &'static str = "../input/input.txt";
const PARSER_REGEX: &'static str = "(?:(\\d+))x(?:(\\d+))x(?:(\\d+))";

fn main() -> Result<(), Box<dyn std::error::Error>> {
    let regex = Regex::new(PARSER_REGEX)?;
    let mut total: usize = 0;
    let mut ribbon: usize = 0;

    // --- Read and parse the input file ---
    let file = File::open(INPUT_FILE)?;
    for line in BufReader::new(file).lines() {
        let line = line.unwrap();
//println!("{}", &line);
        for cap in regex.captures_iter(&line) {
            let l: usize = cap[1].parse()?;
            let w: usize = cap[2].parse()?;
            let h: usize = cap[3].parse()?;
            let a = l * w;
            let b = w * h;
            let c = l * h;
            let m = min(a, min(b, c));
            total += 2 * a + 2 * b + 2 * c + m;
//println!("{}x{}x{} - {}x{}x{}x{} - {}", l, w, h, a, b, c, m, total);

            let m = max(l, max(w, h));
            ribbon += 2 * (l + w + h - m) + l * w * h;
        }
    }

    // --- Puzzle 1 ---
    println!("1. Total square feet of wrapping paper: {}", total);

    // --- Puzzle 2 ---
    println!("2. Total feet of ribbon: {}", ribbon);

    Ok(())
}

File: 2015/02/javascript/main.mjs
import fs from "node:fs";
import readline from "node:readline";

// --- Constants ---
const INPUT_FILE = "../input/input.txt";
const REGEX = /(\d+)x(\d+)x(\d+)/

// --- Variables ---
let total = 0;
let ribbon = 0;

// --- Read and parse the input file ---
const stream = fs.createReadStream(INPUT_FILE);
const reader = readline.createInterface({ input: stream, crlDelay: Infinity });
reader.on("line", line => {
	let cap = line.match(REGEX);
	let l = Number(cap[1]);
	let w = Number(cap[2]);
	let h = Number(cap[3]);
	let a = l * w;
	let b = w * h;
	let c = l * h;
	let m = Math.min(a, Math.min(b, c));
	total += 2 * a + 2 * b + 2 * c + m;

	m = Math.max(l, Math.max(w, h));
	ribbon += 2 * (l + w + h - m) + l * w * h;
});
await new Promise(res => reader.once("close", res));

// --- Puzzle 1 ---
console.log("1. Total square feet of wrapping paper: ", total.toLocaleString());

// --- Puzzle 2 ---
console.log("2. total feet of ribbon: ", ribbon.toLocaleString());
